cedar-policy-validator/src/custom_schema/err.rs

```diff
@@ -0,0 +1,322 @@
+}
+
+impl AsRef<Vec<ParseError>> for ParseErrors {
+    fn as_ref(&self) -> &Vec<ParseError> {
+        &self.0
+    }
+}
+
+impl AsMut<Vec<ParseError>> for ParseErrors {
+    fn as_mut(&mut self) -> &mut Vec<ParseError> {
+        &mut self.0
+    }
+}
+
+impl AsRef<[ParseError]> for ParseErrors {
+    fn as_ref(&self) -> &[ParseError] {
+        self.0.as_ref()
+    }
+}
+
+impl AsMut<[ParseError]> for ParseErrors {
+    fn as_mut(&mut self) -> &mut [ParseError] {
+        self.0.as_mut()
+    }
+}
+
+impl Deref for ParseErrors {
+    type Target = Vec<ParseError>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for ParseErrors {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl<T: Into<ParseError>> From<T> for ParseErrors {
+    fn from(err: T) -> Self {
+        vec![err.into()].into()
+    }
+}
+
+impl From<Vec<ParseError>> for ParseErrors {
+    fn from(errs: Vec<ParseError>) -> Self {
+        ParseErrors(errs)
+    }
+}
+
+impl<T: Into<ParseError>> FromIterator<T> for ParseErrors {
+    fn from_iter<I: IntoIterator<Item = T>>(errs: I) -> Self {
+        ParseErrors(errs.into_iter().map(Into::into).collect())
+    }
+}
+
+impl<T: Into<ParseError>> Extend<T> for ParseErrors {
+    fn extend<I: IntoIterator<Item = T>>(&mut self, iter: I) {
+        self.0.extend(iter.into_iter().map(Into::into))
+    }
+}
+impl IntoIterator for ParseErrors {
+    type Item = ParseError;
+    type IntoIter = std::vec::IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+impl<'a> IntoIterator for &'a ParseErrors {
+    type Item = &'a ParseError;
+    type IntoIter = std::slice::Iter<'a, ParseError>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.iter()
+    }
+}
+
+impl<'a> IntoIterator for &'a mut ParseErrors {
+    type Item = &'a mut ParseError;
+    type IntoIter = std::slice::IterMut<'a, ParseError>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.iter_mut()
+    }
+}
+
+impl Display for ParseErrors {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self.first() {
+            Some(first_err) => write!(f, "{first_err}"), // intentionally showing only the first error; see cedar-policy/cedar#326
+            None => write!(f, "{}", Self::DESCRIPTION_IF_EMPTY),
+        }
+    }
+}
+
+impl std::error::Error for ParseErrors {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        self.first().and_then(std::error::Error::source)
+    }
+}
+
+// Except for `.related()`, everything else is forwarded to the first error, if it is present.
+// This ensures that users who only use `Display`, `.code()`, `.labels()` etc, still get rich
+// information for the first error, even if they don't realize there are multiple errors here.
+// See cedar-policy/cedar#326.
+impl Diagnostic for ParseErrors {
```

Comment:
this whole file has a lot of duplication with the Core `ParseErrors` code.  I wonder if we can factor some of it out